Core/include/Acts/EventData/TrackProxy.hpp

```diff
@@ -619,22 +541,127 @@ class TrackProxy {
-  auto& container() {
-    return *m_container;
+  auto innermostTrackState() {
+    using proxy_t = decltype(m_container->trackStateContainer().getTrackState(
+        std::declval<IndexType>()));
+
+    IndexType stem = component<IndexType>(hashString("stemIndex"));
+    if (stem == kInvalid) {
+      return std::optional<proxy_t>{};
+    } else {
+      return std::optional<proxy_t>{
+          m_container->trackStateContainer().getTrackState(stem)};
+    }
+  }
+
+  /// Get a range over the track states of this track. Return value is
+  /// compatible with range based for loop. Const version
+  /// @note This range is from the outside inwards!
+  /// @return Track state range to iterate over
+  auto trackStatesReversed() const {
+    return m_container->reverseTrackStateRange(m_index);
+  }
+
+  /// Get a range over the track states of this track. Return value is
+  /// compatible with range based for loop. Mutable version
+  /// @note Only available if the track proxy is not read-only
+  /// @note This range is from the outside inwards!
+  /// @return Track state range to iterate over
+  template <bool RO = ReadOnly, typename = std::enable_if_t<!RO>>
+  auto trackStatesReversed() {
+    return m_container->reverseTrackStateRange(m_index);
+  }
+
+  /// Get a range over the track states of this track. Return value is
+  /// compatible with range based for loop. This overload returns a const-only
+  /// track state range, which means you cannot modify the track states obtained
+  /// in the iteration.
+  /// @note This range is from the inside out!
+  /// @warning This access direction is only possible if the track states are
+  ///          **forward-linked**.
+  /// @return Track state range to iterate over
+  auto trackStates() const {
+    return m_container->forwardTrackStateRange(m_index);
+  }
+
+  /// Get a range over the track states of this track.
+  /// Return value is compatible with range based for loop.
+  /// This overload returns a mutable track state range, which means you
+  /// can modify the track states obtained in the iteration.
+  /// @note Only available if the track proxy is not read-only
+  /// @note This range is from the inside out!
+  /// @warning This access direction is only possible if the track states are
+  ///          **forward-linked**.
+  /// @return Track state range to iterate over
+  template <bool RO = ReadOnly, typename = std::enable_if_t<!RO>>
+  auto trackStates() {
+    return m_container->forwardTrackStateRange(m_index);
+  }
+
+  /// @}
+
+  /// @anchor track_proxy_track_state_manipulation
+  /// @name TrackProxy track state manipulation
+  /// Methods that manipulate the track states of a track represented by @c TrackProxy.
+  /// @{
+
+  /// Forward connect a track.
+  /// This means setting indices from the inside out on all track states.
+  /// @note Only available if the track proxy is not read-only
+  template <bool RO = ReadOnly, typename = std::enable_if_t<!RO>>
+  void linkForward() {
```

Comment:
branching track states will overwrite the `next` pointer right?

Reply:
Exactly. Which ever one you link last wins.